src/background.c

```diff
@@ -409,6 +474,372 @@ background_class_init (BackgroundClass *klass)
+
+  handles = g_ptr_array_new_with_free_func (g_free);
+
+  G_LOCK (applications);
+  g_hash_table_iter_init (&iter, applications);
+  while (g_hash_table_iter_next (&iter, (gpointer *)&app_id, (gpointer *)&handle))
+    {
+      gboolean found = FALSE;
+      for (j = 0; j < apps->len && !found; j++)
+        {
+          FlatpakInstance *app = g_ptr_array_index (apps, j);
+          found = g_strcmp0 (app_id, flatpak_instance_get_app (app));
+        }
+      if (!found)
+        {
+          g_hash_table_iter_remove (&iter);
+          if (handle)
+            g_ptr_array_add (handles, g_strdup (handle));
+        }
+    }
+  G_UNLOCK (applications);
+
+  for (j = 0; j < handles->len; j++)
+    { 
+      const char *handle = g_ptr_array_index (handles, j);
+      close_notification (handle);
+    }
+}
+
+/*
+ * Returns whether the @app_id was found in the
+ * table of background apps, and if so, sets
+ * @value to the value found for it.
+ */
+static gboolean
+lookup_background_app (const char *app_id,
+                       char **value)
+{
+  gboolean res;
+  char *orig_key;
+  char *orig_val;
+
+  G_LOCK (applications);
+  res = g_hash_table_lookup_extended (applications, app_id, (gpointer *)&orig_key, (gpointer *)&orig_val);
+  if (res) 
+    *value = g_strdup (orig_val);
+  G_UNLOCK (applications);
+
+  return res;
+}
+
+static void
+add_background_app (const char *app_id,
+                    const char *handle)
+{
+  G_LOCK (applications);
+  g_hash_table_insert (applications, g_strdup (app_id), g_strdup (handle));
+  G_UNLOCK (applications);
+}
+
+static void
+remove_background_app (const char *app_id)
+{
+  g_autofree char *handle = NULL;
+
+  G_LOCK (applications);
+  handle = g_strdup (g_hash_table_lookup (applications, app_id));
+  g_hash_table_remove (applications, app_id);
+  G_UNLOCK (applications);
+
+  if (handle)
+    close_notification (handle);
+}
+
+static char *
+flatpak_instance_get_display_name (FlatpakInstance *instance)
+{
+  const char *app_id = flatpak_instance_get_app (instance);
+  if (app_id[0] != 0)
+    {
+      g_autofree char *desktop_id = NULL;
+      g_autoptr(GAppInfo) info = NULL;
+
+      desktop_id = g_strconcat (app_id, ".desktop", NULL);
+      info = (GAppInfo*)g_desktop_app_info_new (desktop_id);
+
+      if (info)
+        return g_strdup (g_app_info_get_display_name (info));
+    }
+
+  return g_strdup (app_id);
+}
+
+static FlatpakInstance *
+find_instance (const char *app_id)
+{
+  g_autoptr(GPtrArray) instances = NULL;
+  int i;
+
+  instances = flatpak_instance_get_all ();
+  for (i = 0; i < instances->len; i++)
+    {
+      FlatpakInstance *inst = g_ptr_array_index (instances, i);
+
+      if (g_str_equal (flatpak_instance_get_app (inst), app_id))
+        return g_object_ref (inst);
+    }
+
+  return NULL;
+}
+
+static void
+kill_app (const char *app_id)
+{
+  g_autoptr(FlatpakInstance) instance = NULL;
+
+  g_debug ("Killing app %s", app_id);
+
+  instance = find_instance (app_id);
```

Comment:
What if there are multiple instances? This seems to be a way to escape this. We should kill them all.